src/wp-admin/includes/image.php

```diff
@@ -449,24 +571,168 @@ function _wp_make_subsizes( $new_sizes, $file, $image_meta, $attachment_id ) {
+					$primary_size = isset( $image_meta['sizes'][ $new_size_name ]['filesize'] ) ? $image_meta['sizes'][ $new_size_name ]['filesize'] : 0;
+
+					if ( $new_size && $new_size >= $primary_size ) {
+						wp_delete_file( dirname( $file ) . '/' . $new_size_meta['file'] );
+						continue;
+					}
+				}
+				if ( ! isset( $image_meta['sizes'][ $new_size_name ]['sources'] ) ) {
+					$image_meta['sizes'][ $new_size_name ]['sources'] = array();
+				}
+				$image_meta['sizes'][ $new_size_name ]['sources'][ $mime_type ] = _wp_get_sources_from_meta( $new_size_meta );
 				wp_update_attachment_metadata( $attachment_id, $image_meta );
 			}
 		}
 	} else {
 		// Fall back to `$editor->multi_resize()`.
-		$created_sizes = $editor->multi_resize( $new_sizes );
+		$created_sizes = $editor->multi_resize( $new_sizes, true );
 
 		if ( ! empty( $created_sizes ) ) {
-			$image_meta['sizes'] = array_merge( $image_meta['sizes'], $created_sizes );
+			foreach ( $created_sizes as $created_size_name => $created_size_meta ) {
+
+				// Primary mime type is set in 'sizes' array.
+				if ( ! isset( $image_meta['sizes'][ $created_size_name ] ) ) {
+					$image_meta['sizes'][ $created_size_name ] = $created_size_meta;
+				} else {
+					// Remove any newly generated images that are larger than the primary mime type.
+					$new_size     = isset( $created_size_meta['filesize'] ) ? $created_size_meta['filesize'] : 0;
+					$primary_size = isset( $image_meta['sizes'][ $created_size_name ]['filesize'] ) ? $image_meta['sizes'][ $created_size_name ]['filesize'] : 0;
+
+					if ( $new_size && $new_size >= $primary_size ) {
+						wp_delete_file( dirname( $file ) . '/' . $created_size_meta['file'] );
+						continue;
+					}
+				}
+				if ( ! isset( $image_meta['sizes'][ $created_size_name ]['sources'] ) ) {
+					$image_meta['sizes'][ $created_size_name ]['sources'] = array();
+				}
+				$image_meta['sizes'][ $created_size_name ]['sources'][ $mime_type ] = _wp_get_sources_from_meta( $new_size_meta );
+			}
+			wp_update_attachment_metadata( $attachment_id, $image_meta );
+		}
+	}
+
+	return $image_meta;
+}
+
+/**
+ * Get the list of image sizes that support multi-mime type output.
+ *
+ * @since 6.1.0
+ *
+ * @param int $attachment_id Attachment ID.
+ * @return array the list of size names that support multi-mime type output.
+ */
+function _wp_multi_mime_get_supported_sizes( $attachment_id ) {
+	// For WordPress 6.1, output WebP by default for core/core theme sizes.
+	$default_core_sizes = array(
+		'thumbnail',
+		'medium',
+		'medium_large',
+		'large',
+		// 2x medium_large size.
+		'1536x1536',
+		// 2x large size.
+		'2048x2048',
+		// Twentyeleven theme.
+		'large-feature',
+		'small-feature',
+		// Twentyfourteen theme.
+		'twentyfourteen-full-width',
+		// Twentyseventeen theme.
+		'twentyseventeen-featured-image',
+		'twentyseventeen-thumbnail-avatar',
+		// Twentytwenty theme.
+		'twentytwenty-fullscreen',
```

Comment:
This should only have the core sizes which are not using `add_image_size()`. Those that use `add_image_size()` should be adjusted later at the place where they call that function - but I think that should be a separate ticket/PR to not make this one even more complex.
```suggestion
		'thumbnail',
		'medium',
		'medium_large',
		'large',
		'post-thumbnail',
```